pkg/skaffold/kubernetes/portforward/kubectl_forwarder.go

```diff
@@ -107,18 +127,29 @@ func (*KubectlForwarder) Terminate(p *portForwardEntry) {
 // Monitor monitors the logs for a kubectl port forward command
 // If it sees an error, it calls back to the EntryManager to
 // retry the entire port forward operation.
-func (*KubectlForwarder) Monitor(p *portForwardEntry, retryFunc func()) {
+func (*KubectlForwarder) monitorErrorLogs(ctx context.Context, buf *bytes.Buffer, cmd *exec.Cmd, p *portForwardEntry) {
 	for {
-		time.Sleep(1 * time.Second)
-		s, _ := p.logBuffer.ReadString(byte('\n'))
-		if s != "" {
-			logrus.Tracef("[port-forward] %s", s)
-			if strings.Contains(s, "error forwarding port") || strings.Contains(s, "unable to forward") {
-				// kubectl is having an error. retry the command
-				logrus.Infof("retrying kubectl port-forward due to error: %s", s)
-				go retryFunc()
-				return
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			time.Sleep(1 * time.Second)
+			s, _ := buf.ReadString(byte('\n'))
+			if s != "" {
+				logrus.Tracef("[port-forward] %s", s)
+
+				if strings.Contains(s, "error forwarding port") ||
+					strings.Contains(s, "unable to forward") ||
+					strings.Contains(s, "error upgrading connection") {
```

Comment:
Would it make sense to log a warning or something so the user knows the retry is happening and why?

Reply:
it is there in trace mode

Reply:
I keep thinking about what you mentioned: "error upgrading connection" is from port forwarding in client-go: https://github.com/kubernetes/client-go/blob/master/tools/portforward/portforward.go#L194 - I _think_ it is fine to retry on it...

Reply:
I'm pretty sure Cloud Code doesn't show trace — it would be good for users to at least be able to discover that there was a failure.  

I think restarting on error makes a lot of sense: it's essentially turning `kubectl port-forward` into a single-use attempt.  Perhaps this code could look for the more general error header (`portforward.go:.*error occurred`) and log the error but suppress when one of these strings (known issues)?